src/index.tsx

```diff
@@ -16,50 +16,97 @@ import {
 } from "./state";
 import { initialState } from "./state/selection/reducer";
 import { UrlState } from "./util";
-import { URLSearchParam } from "./util/UrlState";
+import { URLSearchParam, URLSearchParamMap } from "./util/UrlState";
 import "./style.css";
 
-const pram = new Pram(createHistory());
+const validateParams = (obj: Record<string, any>): URLSearchParamMap => {
+    const result: URLSearchParamMap = {};
+    for (const key of Object.keys(obj)) {
+        const value = obj[key];
+        if (typeof value === "string" || (Array.isArray(value) && typeof value[0] === "string")) {
+            result[key] = value;
+        }
+    }
+    return result;
+};
+
+class URLParams {
+    constructor(private history: History) {}
+
+    getParams(): URLSearchParamMap {
+        return validateParams(parse(this.history.location.search, { ignoreQueryPrefix: true }));
+    }
+
+    pushParams(params: Record<string, any>) {
+        this.history.push({ ...this.history.location, search: stringify(params) });
+    }
+
+    pushParam(key: string, value: any) {
+        this.pushParams({ ...this.getParams(), [key]: value });
+    }
+
+    replaceParams(params: Record<string, any>) {
+        this.history.replace({ ...this.history.location, search: stringify(params) });
+    }
+
+    onParamRemoved(key: string, callback: () => void) {
+        let hasParam = false;
+        this.history.listen((location) => {
+            const params = parse(location.search, { ignoreQueryPrefix: true });
+            if (params[key]) {
+                hasParam = true;
+            } else if (hasParam) {
+                callback();
+                hasParam = false;
+            }
+        })
+    }
+}
+
+const params = new URLParams(createBrowserHistory());
 const initialSelections = {
     ...initialState,
-    ...UrlState.toAppState(pram.getParams())
+    ...UrlState.toAppState(params.getParams())
 }
 const store = createReduxStore({ selection: initialSelections });
 
-store.dispatch(selection.actions.syncStateWithURL(pram.getParams()));
+store.dispatch(selection.actions.syncStateWithURL(params.getParams()));
 
 // when the dataset query param changes, check if it's been removed
-pram.onChange("dataset", function notifyHostFrameOfURLChange(dataset) {
-    if (!dataset) {
-        // used the back button to get back to the landing page
-        // UrlState doesn't sync actions if a query param doesn't exist, 
-        // so clearing it out here 
-        // NOTE: order matters, the selections need to be cleared out before the 
-        // dataset metadata is cleared
-        store.dispatch(selection.actions.clearDataset());
-        store.dispatch(metadata.actions.clearDatasetValues());
-    }
+params.onParamRemoved("dataset", () => {
+    // used the back button to get back to the landing page
```

Comment:
why did the dataset argument go away here?

Reply:
The original call was something like `onParamChanged`, but the only use of the parameter in the actual body of the function was a null check. I figured that instead of keeping `pram`'s full-featured change detection mechanism, which required maintaining an entire second copy of the URL parameters, I could narrow the listener's job to `onParamRemoved` and just keep one `boolean` for whether the parameter exists or not.